### DISCUSSION TITLE
how to run an integrated test that requires a reactive transaction manager using coroutines?

### DISCUSSION THREAD

**Discussion by lucasgust** [micronaut-projects/micronaut-data#2451] [Q&A] [answered]
I'm developing an app using Micronaut, Kotlin (with coroutines) and R2DBC (here's the [code](https://github.com/lucasgust/txwithr2dbc)) and after upgrading Micronaut version (to v4) I'm not able to make integrated tests work due to reactive transaction context.

good news is that this error is happening only in test environment. in my local environment it is working perfectly.

**Comment by dstepanov** [comment, marked as answer]
Avoid combining coroutines and R2DBC. In the future, we should have some Coroutine TX operations to avoid R2dbc code in Kotlin.

For now, you can create:
```kotlin
@Singleton
open class KotlinTransactional {

    @Transactional
    open suspend fun <T> inTx(tx: suspend () -> T): T {
        return tx.invoke()
    }

}
```
And use it:
```kotlin
        beforeEach {
            runBlocking { // Might not be needed
                kotlinTransactional.inTx {
                    employeeDataAccessPort.deleteAll()
                }
            }
        }
```

**Reply by lucasgust** [reply]
@dstepanov thanks for your response! it's working now.

> Avoid combining coroutines and R2DBC

omg, really? 😢 I struggled for days to make this happen. after V4, I made a several stress tests and everything is working perfectly with R2DBC, coroutines and transactions. I suppose it's because I'm using `CoroutineCrudRepository` instead of `ReactorCrudRepository`.

**Reply by dstepanov** [reply]
Unfortunately it’s really complicated to combine those technologies, the main problem is to propagate the TX context

**Reply by lucasgust** [reply]
I understand and I saw that using older versions because results were very unstable. however V4 seems to be solid, everything is working fine (not even a failure so far), so I thought I could maintain R2DBC.
